sstables/sstables.cc

```diff
@@ -513,23 +513,32 @@ void write(sstable_version_types v, file_writer& out, const disk_set_of_tagged_u
 future<> parse(const schema& schema, sstable_version_types v, random_access_reader& in, summary& s) {
     using pos_type = typename decltype(summary::positions)::value_type;
 
+    constexpr size_t max_chunk_size = 1024 * 128;
+    static_assert(max_chunk_size % sizeof(pos_type) == 0,
+        "parse() assumes that the maximum chunk size (128 KiB) is a multiple of pos_type");
+
     co_await parse(schema, v, in, s.header.min_index_interval,
                      s.header.size,
                      s.header.memory_size,
                      s.header.sampling_level,
                      s.header.size_at_full_sampling);
-    auto buf = co_await in.read_exactly(s.header.size * sizeof(pos_type));
     auto len = s.header.size * sizeof(pos_type);
-    check_buf_size(buf, len);
+    auto fragmented_buffers = co_await in.read_exactly_fragmented(len);
 
-    // Positions are encoded in little-endian.
-    auto b = buf.get();
     s.positions.reserve(s.header.size + 1);
-    while (s.positions.size() != s.header.size) {
-        s.positions.push_back(seastar::read_le<pos_type>(b));
-        b += sizeof(pos_type);
-        co_await coroutine::maybe_yield();
+    for (auto& buf : fragmented_buffers) {
+        check_buf_size(buf, std::min(len, max_chunk_size));
+        const auto count_in_curr_chunk = buf.size() / sizeof(pos_type);
+        auto b = buf.get();
+        for (size_t i = 0; i < count_in_curr_chunk; ++i) {
```

Comment:
Why do we need to iterate using an index if we simply compare b to an end pointer: buf.get + buf.size()?

Reply:
Fixed in V2.